ngx_http_upstream_jdomain.c

```diff
@@ -443,17 +456,31 @@ ngx_http_upstream_jdomain_handler(ngx_resolver_ctx_t *ctx)
 
 	ngx_log_debug3(NGX_LOG_DEBUG_CORE, r->log, 0, "upstream_jdomain: \"%V\" resolved state(%i: %s)", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state));
 
-	if (ctx->state || ctx->naddrs == 0) {
-		if (urcf->fallback_addr.sockaddr == NULL) {
-			ngx_log_error(NGX_LOG_ERR, r->log, 0, "upstream_jdomain: resolver failed, \"%V\" (%i: %s)", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state));
-
-			goto end;
-		} else {
-			ngx_log_error(NGX_LOG_WARN, r->log, 0, "upstream_jdomain: resolver failed, \"%V\" (%i: %s), using fallback address \"%V\"", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state), &urcf->fallback_addr.name);
-			ctx->addrs = (ngx_resolver_addr_t *)&urcf->fallback_addr;
-			port = urcf->fallback_port;
-			ctx->naddrs = 1;
-		}
+	if (ctx->state || ctx->naddrs == 0) {	
+	    if (urcf->fallback_addr.sockaddr != NULL ) {
+	        if (urcf->fallback_strict == 1 ) {
+                ngx_log_error(NGX_LOG_WARN, r->log, 0, "upstream_jdomain: resolver failed, \"%V\" (%i: %s), strict using fallback address \"%V\"", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state), &urcf->fallback_addr.name);
+                ctx->addrs = (ngx_resolver_addr_t *)&urcf->fallback_addr;
+                port = urcf->fallback_port;
+                ctx->naddrs = 1;
+	        } else {
+                switch (ctx->state) {
+                    case 1:
+                    case 3:
+                        ngx_log_error(NGX_LOG_WARN, r->log, 0, "upstream_jdomain: resolver failed, \"%V\" (%i: %s), non-strict using fallback address \"%V\"", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state), &urcf->fallback_addr.name);
+                        ctx->addrs = (ngx_resolver_addr_t *)&urcf->fallback_addr;
+                        port = urcf->fallback_port;
+                        ctx->naddrs = 1;
+                        break;
+                    default:
+                        ngx_log_error(NGX_LOG_ERR, r->log, 0, "upstream_jdomain: resolver failed without fallback, \"%V\" (%i: %s)", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state));
+                        goto end;    
+                }
+	        }
+	    } else {
+            ngx_log_error(NGX_LOG_ERR, r->log, 0, "upstream_jdomain: resolver failed, \"%V\" (%i: %s)", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state));
+            goto end;
+	    }
```

Comment:
I think maybe it was actually bad advice to use a switch case here, since it makes the logic look a lot more complicated and requires us to duplicate code.

I think it would be more simple to just use a single `if` block here, also I think we should refer to the errors by their constant names instead of numeric values, in case the nginx developers decide to change them.

I think something like this would be better:

```suggestion
		if (urcf->fallback_addr.sockaddr && (urcf->fallback_strict || ctx->state == NGX_RESOLVE_FORMERR || ctx->state == NGX_RESOLVE_NXDOMAIN) {
			ngx_log_error(NGX_LOG_WARN, r->log, 0, "upstream_jdomain: resolver failed, \"%V\" (%i: %s),%s using fallback address \"%V\"", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state), urcf->fallback_strict ? " strictly" : "", &urcf->fallback_addr.name);
			ctx->addrs = (ngx_resolver_addr_t *)&urcf->fallback_addr;
			port = urcf->fallback_port;
			ctx->naddrs = 1;
		} else {
			ngx_log_error(NGX_LOG_ERR, r->log, 0, "upstream_jdomain: resolver failed, \"%V\" (%i: %s)", &ctx->name, ctx->state, ngx_resolver_strerror(ctx->state));
			goto end;
		}
```

The only thing I'm not 100% sure about is the ternary expression I added in the log `urcf->fallback_strict ? " strictly" : ""`, but I'm also not sure how much we even care to know from the log if it is using strict or not. I think we maybe only care if it's using fallback or not.

Reply:
Also note: I'm not sure but you may need to add another `#include` in order to reference `NGX_RESOLVE_FORMERR` and `NGX_RESOLVE_NXDOMAIN`. I don't think so though since we are able to reference functions like `ngx_resolver_strerror`, but it's something to keep in mind in case you see build errors.